pex/orderedset.py

```diff
@@ -10,71 +10,65 @@
 
 from __future__ import absolute_import
 
+from collections import OrderedDict
+
 from pex.compatibility import MutableSet
+from pex.typing import TYPE_CHECKING, Generic
+
+if TYPE_CHECKING:
+    from typing import Any, Iterable, Optional, TypeVar
 
+    _I = TypeVar("_I")
 
-class OrderedSet(MutableSet):
-    KEY, PREV, NEXT = range(3)
 
+class OrderedSet(MutableSet, Generic["_I"]):
     def __init__(self, iterable=None):
-        self.end = end = []
-        end += [None, end, end]  # sentinel node for doubly linked list
-        self.map = {}  # key --> [key, prev, next]
+        # type: (Optional[Iterable[_I]]) -> None
+        self.map = OrderedDict()  # type: OrderedDict[_I, None]
         if iterable is not None:
-            self |= iterable
+            self.update(iterable)
 
     def __len__(self):
+        # type: () -> int
         return len(self.map)
 
     def __contains__(self, key):
+        # type: (Any) -> bool
         return key in self.map
 
     def add(self, key):
-        if key not in self.map:
-            end = self.end
-            curr = end[self.PREV]
-            curr[self.NEXT] = end[self.PREV] = self.map[key] = [key, curr, end]
+        # type: (_I) -> None
+        self.map[key] = None
 
     def update(self, iterable):
+        # type: (Iterable[_I]) -> None
         for key in iterable:
             self.add(key)
 
     def discard(self, key):
-        if key in self.map:
-            key, prev, next = self.map.pop(key)
-            prev[self.NEXT] = next
-            next[self.PREV] = prev
+        # type: (_I) -> None
+        self.map.pop(key, None)
 
     def __iter__(self):
-        end = self.end
-        curr = end[self.NEXT]
-        while curr is not end:
-            yield curr[self.KEY]
-            curr = curr[self.NEXT]
+        return iter(self.map)
 
     def __reversed__(self):
-        end = self.end
-        curr = end[self.PREV]
-        while curr is not end:
-            yield curr[self.KEY]
-            curr = curr[self.PREV]
+        return reversed(self.map)
 
     def pop(self, last=True):
         if not self:
             raise KeyError("set is empty")
-        key = next(reversed(self)) if last else next(iter(self))
-        self.discard(key)
+        key, _ = self.map.popitem(last=last)
         return key
 
     def __repr__(self):
         if not self:
-            return "%s()" % (self.__class__.__name__,)
-        return "%s(%r)" % (self.__class__.__name__, list(self))
+            return "{}()".format(
+                self.__class__.__name__,
+            )
+        return "{}({!r})".format(self.__class__.__name__, list(self))
 
     def __eq__(self, other):
         if isinstance(other, OrderedSet):
             return len(self) == len(other) and list(self) == list(other)
```

Comment:
Directly comparing `self.map` should be cheaper than converting to a list, as `OrderedDict` It takes ordering into account. Downside might be that Python3's dict does _not_ take ordering into account in eq, so would need a comment to fix that when Python2 support is dropped.

Reply:
Fixed. I bit the bullet and re-implemented `__eq__` completely taking away its ability to compare `OrderedSet([1, 2, 3]) == {1, 2, 3}` which was all of strange, wrong and unused.